### PR TITLE
Change the type of differenceR to include a mandatory base shape

### LINKED ISSUE #293: exportSTL containing a difference with only a single element doesn't produce any output
```haskell
main = do
  writeSTL 2 "/tmp/test.stl" $
    difference [ sphere 3 ] 
```

doesn't result in a sphere; instead it produces a mesh with no triangles.

Instead, `difference (x : [])` should be equal to `x`.

### PR DESCRIPTION
**Pull request by isovector**
tl;dr: changes the type of `differenceR` from `Object obj vec => R -> [obj] -> obj` to `Object obj vec => R -> obj -> [obj] -> obj`.

The first element of the list argument to `differenceR` is special, and
acts as the shape from which the others are subtracted. The previous
behavior of exporting `differenceR r []` (or even `differenceR [x]`) was
to crash due to `head` and `tail` being partial functions.

From this analysis, it's clear that the "shape to remove from" is
a *mandatory* argument to `differenceR`, and as such should be
explicitly moved out of the list args.

Fixes #293

**Note:** I don't understand how the scad parser should deal with this, so I just hammered this in there. Please take a look (right now I'm just calling `error`.)

### CODE REVIEW

**Comment by isovector** [comment]
A cursory glance at the parsing machinery didn't suggest any way of communicating errors to the user. @julialongtin any pointers to how this should work?

**Comment by isovector** [comment]
@julialongtin PTAL

**Review by julialongtin — APPROVED** [review]
